File: parser.c
#include "parser.h"
#include "error.h"
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MSG_LEN 8

int alloc_cur = 0;

ASTNode *alloc_node() {
  if (alloc_cur == MAX_NODE) {
    error("can't allocate ASTNode, max: %d", MAX_NODE);
  }
  return &ast_node_pool[alloc_cur++];
}

void init_parser(Parser *p, char *source) {
  p->source = source;
  p->len = strlen(source);
  p->cur = 0;
}

static void __attribute__((noreturn)) parse_error(Parser *p, char *expected) {
  char err_msg[MSG_LEN + 1];
  memcpy(err_msg, &p->source[p->cur], MSG_LEN);
  err_msg[MSG_LEN] = '\0';
  error(
      "parse token error at %d, expected %s, but got unexpected token: '%s'\n",
      p->cur, expected, err_msg);
}

static char next_c(Parser *p) {
  if (p->cur == p->len)
    return '\0';
  while (p->source[p->cur] == ' ') {
    p->cur++;
  }
  return p->source[p->cur];
}

static int match_str(Parser *p, char *str, int consume) {
  next_c(p);
  int len = strlen(str);
  int ret = memcmp(&p->source[p->cur], str, len);
  if (ret == 0) {
    p->cur += len;
  } else if (consume) {
    parse_error(p, str);
  }
  return ret == 0;
}

static Token read_token(Parser *p) {
  char c = next_c(p);
  switch (c) {
  case 'p':
    if (match_str(p, "program", 0))
      return Program;
  case 'l':
    if (match_str(p, "let", 0))
      return Let;
  case '+':
    p->cur++;
    return Add;
  case '-':
    p->cur++;
    return Neg;
  case '0':
  case '1':
  case '2':
  case '3':
  case '4':
  case '5':
  case '6':
  case '7':
  case '8':
  case '9':
    return Fixnum;
  case 'r':
    if (match_str(p, "read", 0))
      return Read;
  default:
    return Var;
  }
}

static int read_fixnum(Parser *p) {
  int cur = p->cur;
  int i = 0;
  int n = 0;
  while (1) {
    int c = p->source[p->cur++];
    if (c > '9' || c < '0') {
      break;
    }
    n = n * 10 + i * (c - '0');
    i++;
  }
  p->cur--;
  if (i == 0) {
    parse_error(p, "fixnum");
  }
  return n;
}

static char *read_var(Parser *p) {
  int cur = p->cur;
  int i = 0;
  char *var = malloc(10);
  while (i < 9) {
    int c = p->source[p->cur++];
    if (c > 'z' || c < 'a') {
      break;
    }
    var[i] = c;
    i++;
  }
  var[i] = '\0';
  p->cur--;
  if (i == 0) {
    parse_error(p, "var");
  }
  return var;
}

static ASTNode *read_exp(Parser *p) {
  int in_paren = match_str(p, "(", 0);
  Token token = read_token(p);
  ASTNode *n = alloc_node();
  n->token = token;
  switch (token) {
  case Add:
    n->lhs = read_exp(p);
    n->rhs = read_exp(p);
    break;
  case Neg:
    n->lhs = read_exp(p);
    break;
  case Fixnum:
    n->value = read_fixnum(p);
    break;
  case Read:
    break;
  case Let:
    match_str(p, "(", 1);
    match_str(p, "[", 1);
    ASTNode *val_n = alloc_node();
    val_n->value = (size_t)read_var(p);
    next_c(p);
    n->value = read_fixnum(p);
    n->lhs = val_n;
    match_str(p, "]", 1);
    match_str(p, ")", 1);
    n->rhs = read_exp(p);
    break;
  case Var:
    n->value = (size_t)read_var(p);
    break;
  default:
    parse_error(p, "token");
  }
  if (in_paren)
    match_str(p, ")", 1);
  return n;
}

ASTNode *parse_program(Parser *p) { return read_exp(p); }

File: src/parser.rs
use crate::ast::{Node, Token};
use std::iter::FromIterator;

pub struct Parser {
    source: Vec<char>,
    cur: usize,
}

impl Parser {
    pub fn new(source: Vec<char>) -> Self {
        Parser { source, cur: 0 }
    }

    fn next_char(&mut self) -> Option<char> {
        if self.cur == self.source.len() {
            return None;
        }
        while self.cur < self.source.len() && self.source[self.cur] == ' ' {
            self.cur += 1;
        }
        if self.cur == self.source.len() {
            return None;
        }
        Some(self.source[self.cur])
    }

    fn match_str(&mut self, expected: &str) -> Result<(), String> {
        self.next_char();
        let end_at = std::cmp::min(self.source.len(), self.cur + expected.len());
        let actual = &self.source[self.cur..end_at];
        let matched = actual == &expected.chars().collect::<Vec<_>>()[..];
        if matched {
            self.cur += expected.len();
            Ok(())
        } else {
            Err(String::from_iter(actual.iter()))
        }
    }

    fn expect_str(&mut self, expected: &str) {
        if let Err(actual) = self.match_str(expected) {
            panic!(
                "parse token error at {}, expected {}, but got unexpected token: '{}'",
                self.cur, expected, actual
            );
        }
    }

    fn read_fixnum(&mut self) -> Option<isize> {
        let prev_cur = self.cur;
        let mut num = 0;
        loop {
            let chr = self.source[self.cur];
            if chr > '9' || chr < '0' {
                break;
            }
            num = num * 10 + chr.to_digit(10).unwrap() as isize;
            self.cur += 1;
        }
        if prev_cur == self.cur {
            return None;
        }
        Some(num)
    }

    fn read_var(&mut self) -> Option<String> {
        let prev_cur = self.cur;
        let mut var = String::new();
        loop {
            let chr = self.source[self.cur];
            if chr > 'z' || chr < 'a' {
                break;
            }
            self.cur += 1;
            var.push(chr);
        }
        if prev_cur == self.cur {
            return None;
        }
        Some(var)
    }

    fn read_token(&mut self) -> Option<Token> {
        use Token::*;

        let chr = match self.next_char() {
            Some(chr) => chr,
            None => return None,
        };
        let token = match chr {
            'p' if self.match_str("program").is_ok() => Program,
            'l' if self.match_str("let").is_ok() => Let,
            '+' => {
                self.cur += 1;
                Add
            }
            '-' => {
                self.cur += 1;
                Neg
            }
            'n' if self.match_str("not").is_ok() => Not,
            't' if self.match_str("true").is_ok() => True,
            'f' if self.match_str("false").is_ok() => False,
            '=' if self.match_str("==").is_ok() => Eq,
            '<' if self.match_str("<=").is_ok() => Lte,
            '<' => {
                self.cur += 1;
                Lt
            }
            '>' if self.match_str(">=").is_ok() => Gte,
            '>' => {
                self.cur += 1;
                Gt
            }
            '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => Fixnum,
            'r' if self.match_str("read").is_ok() => Read,
            _ => Var,
        };
        Some(token)
    }

    fn read_exp(&mut self) -> Box<Node> {
        use Token::*;

        let in_paren = self.match_str("(").is_ok();
        let token = self.read_token().expect("token");
        let node = match token {
            Program => Node::Program(self.read_exp()),
            Add => Node::Add(self.read_exp(), self.read_exp()),
            Neg => Node::Neg(self.read_exp()),
            Fixnum => Node::Fixnum(self.read_fixnum().expect("fixnum")),
            Read => Node::Read,
            Let => {
                self.expect_str("(");
                self.expect_str("[");
                let name = self.read_var().expect("var");
                self.next_char();
                let value = self.read_exp();
                match value.as_ref() {
                    Node::Fixnum(_) | Node::True | Node::False => {}
                    v => {
                        panic!("expect value, got {:?}", v);
                    }
                }
                self.expect_str("]");
                self.expect_str(")");
                Node::Let {
                    name,
                    value,
                    exp: self.read_exp(),
                }
            }
            Var => Node::Var(self.read_var().expect("var")),
            True => Node::True,
            False => Node::False,
            Not => Node::Not(self.read_exp()),
            Eq => Node::Eq(self.read_exp(), self.read_exp()),
            Lt => Node::Lt(self.read_exp(), self.read_exp()),
            Lte => Node::Lte(self.read_exp(), self.read_exp()),
            Gt => Node::Gt(self.read_exp(), self.read_exp()),
            Gte => Node::Gte(self.read_exp(), self.read_exp()),
        };
        if in_paren {
            self.expect_str(")");
        }
        return Box::new(node);
    }

    pub fn parse_program(&mut self) -> Box<Node> {
        self.read_exp()
    }
}
